ProcessLib/SourceTerms/VolumetricSourceTermFEM.h

```diff
@@ -0,0 +1,104 @@
+
+namespace ProcessLib
+{
+class VolumetricSourceTermLocalAssemblerInterface
+    : public ProcessLib::LocalAssemblerInterface
+{
+public:
+    VolumetricSourceTermLocalAssemblerInterface() = default;
+
+    virtual void integrate(
+        std::size_t const id,
+        NumLib::LocalToGlobalIndexMap const& source_term_dof_table,
+        double const t, GlobalVector& b) = 0;
+};
+
+const unsigned NUM_NODAL_DOF = 1;
+
+template <typename ShapeFunction, typename IntegrationMethod,
+          unsigned GlobalDim>
+class VolumetricSourceTermLocalAssembler
+    : public VolumetricSourceTermLocalAssemblerInterface
+{
+    using ShapeMatricesType = ShapeMatrixPolicyType<ShapeFunction, GlobalDim>;
+    using ShapeMatrices = typename ShapeMatricesType::ShapeMatrices;
+
+    using LocalAssemblerTraits = ProcessLib::LocalAssemblerTraits<
+        ShapeMatricesType, ShapeFunction::NPOINTS, NUM_NODAL_DOF, GlobalDim>;
+
+    using NodalMatrixType = typename LocalAssemblerTraits::LocalMatrix;
+    using NodalVectorType = typename LocalAssemblerTraits::LocalVector;
+    using GlobalDimVectorType = typename ShapeMatricesType::GlobalDimVectorType;
+
+public:
+    VolumetricSourceTermLocalAssembler(
+        MeshLib::Element const& element,
+        std::size_t const /*local_matrix_size*/,
+        bool is_axially_symmetric,
+        unsigned const integration_order,
+        VolumetricSourceTermData const& process_data)
+        : _process_data(process_data),
+          _integration_method(integration_order),
+          _shape_matrices(initShapeMatrices<ShapeFunction, ShapeMatricesType,
+                                            IntegrationMethod, GlobalDim>(
+              element, is_axially_symmetric, _integration_method))
+    {
+    }
+
+    void integrate(std::size_t const id,
+                   NumLib::LocalToGlobalIndexMap const& source_term_dof_table,
+                   double const t, GlobalVector& b) override
+    {
+        _local_rhs.setZero();
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        SpatialPosition pos;
+        pos.setElementID(id);
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            pos.setIntegrationPoint(ip);
+            auto const& sm = _shape_matrices[ip];
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+            auto const st_val = _process_data.volumetric_source_term(t, pos)[0];
+
+            _local_rhs.noalias() +=
+                sm.N * st_val * sm.detJ * sm.integralMeasure * wp.getWeight();
+        }
+        auto const indices = NumLib::getIndices(id, source_term_dof_table);
+        b.add(indices, _local_rhs);
+    }
+
+private:
+    VolumetricSourceTermData const& _process_data;
+
+    IntegrationMethod const _integration_method;
+    std::vector<ShapeMatrices, Eigen::aligned_allocator<ShapeMatrices>>
+        _shape_matrices;
+    NodalVectorType _local_rhs;
```

Comment:
For dynamic-size shape matrices this probably hs to be initialized with the proper size.